File: TestTasks/src/task2/src/observer_node.cpp
#include "ros/ros.h"
#include <geometry_msgs/PoseStamped.h>
#include <geometry_msgs/PoseArray.h>

using namespace ros;

Publisher pose_array_pub;
std::vector<geometry_msgs::Pose> poses;
Time prev_time;
double interval = 0.1;

void poseCallback(const geometry_msgs::PoseStamped::ConstPtr& msg)
{
  //save one pose per interval
  Time cur_time = Time::now();
  double delta_time = (cur_time - prev_time).nsec / pow(10, 9);
  if (delta_time<interval) return;
  prev_time = cur_time;

  //save pose to array
  geometry_msgs::PoseArray new_msg;
  poses.push_back(msg->pose);
  new_msg.poses = poses;
  new_msg.header.stamp = Time::now();
  new_msg.header.frame_id = "map";
  pose_array_pub.publish(new_msg);
}

int main(int argc, char **argv)
{
  init(argc, argv, "observer_node");
  NodeHandle nh;
  Subscriber pose_sub = nh.subscribe("pose", 1000, poseCallback);
  pose_array_pub = nh.advertise<geometry_msgs::PoseArray>("pose_array", 1000);

  Rate loop_rate(10);
  prev_time = Time::now();

  while (ok())
  {
    spinOnce();
    loop_rate.sleep();
  }

  return 0;
}

File: TestTasks/src/task2/src/observer_node.py
#!/usr/bin/env python

import rospy
from geometry_msgs.msg import PoseStamped
from geometry_msgs.msg import PoseArray

interval = 0.1
poses = []
pose_array_pub = None
prev_time = None

def pose_callback(msg):
    global prev_time, poses, pose_array_pub
    # save one pose per interval
    cur_time = rospy.Time.now()
    delta_time = (cur_time - prev_time).nsecs / float(pow(10, 9))
    if delta_time < interval:
        return
    prev_time = cur_time

    # save pose to array
    poses.append(msg.pose)

    # create pose array message
    new_msg = PoseArray()
    new_msg.poses = poses
    new_msg.header.stamp = cur_time
    new_msg.header.frame_id = "map"
    pose_array_pub.publish(new_msg)

    rospy.loginfo("Runner pose: x=%f, y=%f, z=%f", msg.pose.position.x, msg.pose.position.y, msg.pose.position.z)


def main():
    global prev_time, poses, pose_array_pub
    rospy.init_node('observer_node', anonymous=True)
    rate = rospy.Rate(10)
    prev_time = rospy.Time.now()
    pose_sub = rospy.Subscriber('pose', PoseStamped, pose_callback)
    pose_array_pub = rospy.Publisher("pose_array", PoseArray, queue_size=10)

    while not rospy.is_shutdown():
        rate.sleep()


if __name__ == '__main__':
    main()
